packages/@ember/-internals/routing/lib/system/router.ts

```diff
@@ -47,6 +47,15 @@ import Router, {
 import type { Timer } from 'backburner';
 import { EngineRouteInfo } from './engines';
 import EngineInstance from '@ember/engine/instance';
+import { QueryParams } from 'route-recognizer';
+
+type AnyFn = (...args: any[]) => any;
+
+type MethodNamesOf<T> = {
+  [K in keyof T]: T[K] extends AnyFn ? K : never;
+}[keyof T];
```

Comment:
Not urgent to do before merging, but we should definitely extract these kinds of common type utilities into a shared location to avoid duplicating them/drifting implementations.